In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%cd /content/drive/MyDrive/anlp_notebooks

/content/drive/MyDrive/anlp_notebooks


In [1]:
import torch
print(torch.__version__)
print(torch.cuda.is_available())
print(torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU")

2.10.0+cu128
True
NVIDIA GeForce RTX 5090 Laptop GPU


In [1]:
!pip install gensim

   ---------------------------------------- 0.0/24.4 MB ? eta -:--:--
   ---------------------------------------- 0.0/24.4 MB ? eta -:--:--
   ---------------------------------------- 0.3/24.4 MB ? eta -:--:--
    --------------------------------------- 0.5/24.4 MB 979.3 kB/s eta 0:00:25
   - -------------------------------------- 0.8/24.4 MB 1.0 MB/s eta 0:00:24
   - -------------------------------------- 0.8/24.4 MB 1.0 MB/s eta 0:00:24
   - -------------------------------------- 1.0/24.4 MB 952.8 kB/s eta 0:00:25
   -- ------------------------------------- 1.3/24.4 MB 934.8 kB/s eta 0:00:25
   -- ------------------------------------- 1.3/24.4 MB 934.8 kB/s eta 0:00:25
   -- ------------------------------------- 1.6/24.4 MB 908.6 kB/s eta 0:00:26
   --- ------------------------------------ 1.8/24.4 MB 904.6 kB/s eta 0:00:25
   --- ------------------------------------ 1.8/24.4 MB 904.6 kB/s eta 0:00:25
   --- ------------------------------------ 2.1/24.4 MB 860.7 kB/s eta 0:00:26
   -

In [2]:
from gensim.models import KeyedVectors

mymodel =KeyedVectors.load_word2vec_format("GoogleNews-vectors-negative300/GoogleNews-vectors-negative300.bin",binary=True)

In [3]:
mymodel.similarity('man', 'woman')

np.float32(0.76640123)

In [4]:
mymodel.most_similar(positive=['man'])

[('woman', 0.7664012908935547),
 ('boy', 0.6824871301651001),
 ('teenager', 0.6586930155754089),
 ('teenage_girl', 0.6147903203964233),
 ('girl', 0.5921714305877686),
 ('suspected_purse_snatcher', 0.571636438369751),
 ('robber', 0.5585119128227234),
 ('Robbery_suspect', 0.5584409832954407),
 ('teen_ager', 0.5549196600914001),
 ('men', 0.5489763021469116)]

In [5]:
mymodel['man']

array([ 0.32617188,  0.13085938,  0.03466797, -0.08300781,  0.08984375,
       -0.04125977, -0.19824219,  0.00689697,  0.14355469,  0.0019455 ,
        0.02880859, -0.25      , -0.08398438, -0.15136719, -0.10205078,
        0.04077148, -0.09765625,  0.05932617,  0.02978516, -0.10058594,
       -0.13085938,  0.001297  ,  0.02612305, -0.27148438,  0.06396484,
       -0.19140625, -0.078125  ,  0.25976562,  0.375     , -0.04541016,
        0.16210938,  0.13671875, -0.06396484, -0.02062988, -0.09667969,
        0.25390625,  0.24804688, -0.12695312,  0.07177734,  0.3203125 ,
        0.03149414, -0.03857422,  0.21191406, -0.00811768,  0.22265625,
       -0.13476562, -0.07617188,  0.01049805, -0.05175781,  0.03808594,
       -0.13378906,  0.125     ,  0.0559082 , -0.18261719,  0.08154297,
       -0.08447266, -0.07763672, -0.04345703,  0.08105469, -0.01092529,
        0.17480469,  0.30664062, -0.04321289, -0.01416016,  0.09082031,
       -0.00927734, -0.03442383, -0.11523438,  0.12451172, -0.02

Mikolov et al. (2013) propose the use of an offset vector method in order to answer analogy questions.
For example, if you want to find the concept X which satisfies the analogy “X is to China as London is
to England”, you need to find the concept closest to the point X in the vector space:

$$X = \vec{V}_{\text{China}} + \vec{V}_{\text{London}} - \vec{V}_{\text{England}}$$

In [6]:
mymodel.most_similar(positive=['China','London'],negative=['England'])

[('Beijing', 0.6737731099128723),
 ('Shanghai', 0.6466286778450012),
 ('Beijng', 0.5856549739837646),
 ('Hong_Kong', 0.5709934830665588),
 ('Chinese', 0.563977062702179),
 ('Guangdong', 0.5119545459747314),
 ('Shenzhen', 0.5102902054786682),
 ('Yanqi', 0.5076326727867126),
 ('Nanjing', 0.505686342716217),
 ('Guangzhou', 0.5043154954910278)]

In [7]:
mymodel.most_similar(positive=['China','London'],negative=['England'])[0]

('Beijing', 0.6737731099128723)

The file relations.json contains lists of pairs of words which satisfy some syntactic or semantic
relation. You can load it into a dictionary using the following code

In [8]:
import json
with open('relations.json', 'r') as fp:
  testtuples=json.load(fp)
print(testtuples)

{'gram3-comparative': [['bad', 'worse'], ['big', 'bigger'], ['bright', 'brighter'], ['cheap', 'cheaper'], ['cold', 'colder'], ['cool', 'cooler'], ['deep', 'deeper'], ['easy', 'easier'], ['fast', 'faster'], ['good', 'better'], ['great', 'greater'], ['hard', 'harder'], ['heavy', 'heavier'], ['high', 'higher'], ['hot', 'hotter'], ['large', 'larger'], ['long', 'longer'], ['loud', 'louder'], ['low', 'lower'], ['new', 'newer'], ['old', 'older'], ['quick', 'quicker'], ['safe', 'safer'], ['sharp', 'sharper'], ['short', 'shorter'], ['simple', 'simpler'], ['slow', 'slower'], ['small', 'smaller'], ['smart', 'smarter'], ['strong', 'stronger'], ['tall', 'taller'], ['tight', 'tighter'], ['tough', 'tougher'], ['warm', 'warmer'], ['weak', 'weaker'], ['wide', 'wider'], ['young', 'younger']], 'gram8-plural': [['banana', 'bananas'], ['bird', 'birds'], ['bottle', 'bottles'], ['building', 'buildings'], ['car', 'cars'], ['cat', 'cats'], ['child', 'children'], ['cloud', 'clouds'], ['color', 'colors'], ['comp

## **Tasks**

 Write a function which when given one (capital city, country) training pair can predict the capital
of the other countries in the capital-common-countries list in testtuples

In [9]:
testtuples['capital-common-countries']

[['Athens', 'Greece'],
 ['Baghdad', 'Iraq'],
 ['Bangkok', 'Thailand'],
 ['Beijing', 'China'],
 ['Berlin', 'Germany'],
 ['Bern', 'Switzerland'],
 ['Cairo', 'Egypt'],
 ['Canberra', 'Australia'],
 ['Hanoi', 'Vietnam'],
 ['Havana', 'Cuba'],
 ['Helsinki', 'Finland'],
 ['Islamabad', 'Pakistan'],
 ['Kabul', 'Afghanistan'],
 ['London', 'England'],
 ['Madrid', 'Spain'],
 ['Moscow', 'Russia'],
 ['Oslo', 'Norway'],
 ['Ottawa', 'Canada'],
 ['Paris', 'France'],
 ['Rome', 'Italy'],
 ['Stockholm', 'Sweden'],
 ['Tehran', 'Iran'],
 ['Tokyo', 'Japan']]

In [10]:
testtuples['capital-common-countries']
pairs_count = testtuples['capital-common-countries']
for capital, country in pairs_count:
  print(country)

Greece
Iraq
Thailand
China
Germany
Switzerland
Egypt
Australia
Vietnam
Cuba
Finland
Pakistan
Afghanistan
England
Spain
Russia
Norway
Canada
France
Italy
Sweden
Iran
Japan


In [113]:
def  capital_predictor(train_capital, train_country):
    pairs_count = testtuples['capital-common-countries']
    predictions = {}

    for true_capital, des_country in pairs_count:
        if des_country == train_country:
            continue

        pred = mymodel.most_similar(
            positive=[des_country, train_capital],
            negative=[train_country],
            topn=1
        )

        predictions[des_country] = pred[0]

    return predictions


In [92]:
capital_predictor('Paris', 'France')

{'Greece': [('Athens', 0.6660104990005493)],
 'Iraq': [('Baghdad', 0.6557148694992065)],
 'Thailand': [('Bangkok', 0.7859742045402527)],
 'China': [('Beijing', 0.7495014667510986)],
 'Germany': [('Berlin', 0.7644002437591553)],
 'Switzerland': [('Zurich', 0.6314820647239685)],
 'Egypt': [('Cairo', 0.7625235319137573)],
 'Australia': [('Sydney', 0.7721031308174133)],
 'Vietnam': [('Saigon', 0.6242673993110657)],
 'Cuba': [('Havana', 0.7411302328109741)],
 'Finland': [('Helsinki', 0.7404909729957581)],
 'Pakistan': [('Islamabad', 0.7356792688369751)],
 'Afghanistan': [('Kabul', 0.6445335745811462)],
 'England': [('London', 0.6441212892532349)],
 'Spain': [('Madrid', 0.7571905255317688)],
 'Russia': [('Moscow', 0.8329363465309143)],
 'Norway': [('Oslo', 0.718791127204895)],
 'Canada': [('Toronto', 0.6297447085380554)],
 'Italy': [('Milan', 0.7222140431404114)],
 'Sweden': [('Stockholm', 0.7852394580841064)],
 'Iran': [('Tehran', 0.7437655925750732)],
 'Japan': [('Tokyo', 0.814286172389984

In [53]:
capital_predictor('Kabul', 'Afghanistan')

{'Greece': ('Thessaloniki', 0.6054319143295288),
 'Iraq': ('Baghdad', 0.822655200958252),
 'Thailand': ('Bangkok', 0.7506464123725891),
 'China': ('Beijing', 0.7516018152236938),
 'Germany': ('Cologne', 0.6428322196006775),
 'Switzerland': ('Swiss', 0.6656177639961243),
 'Egypt': ('Cairo', 0.7107338309288025),
 'Australia': ('Sydney', 0.6557939052581787),
 'Vietnam': ('Hanoi', 0.7509632706642151),
 'Cuba': ('Havana', 0.7647624015808105),
 'Finland': ('Helsinki', 0.6525307893753052),
 'Pakistan': ('Islamabad', 0.7911513447761536),
 'England': ('stock_symbol_BNK', 0.5503717064857483),
 'Spain': ('Madrid', 0.6430703401565552),
 'Russia': ('Moscow', 0.7589125037193298),
 'Norway': ('Norwegian', 0.6620277762413025),
 'Canada': ('Canadian', 0.5911223292350769),
 'France': ('Paris', 0.6601777672767639),
 'Italy': ('Italian', 0.6285211443901062),
 'Sweden': ('Stockholm', 0.6646467447280884),
 'Iran': ('Tehran', 0.8056679964065552),
 'Japan': ('Tokyo', 0.7385891675949097)}

In [54]:
capital_predictor('Ottawa', 'Canada')

{'Greece': ('Athens', 0.546576738357544),
 'Iraq': ('Baghdad', 0.6022132039070129),
 'Thailand': ('Bangkok', 0.6644393801689148),
 'China': ('Beijing', 0.6791781187057495),
 'Germany': ('Hamburg', 0.6264716386795044),
 'Switzerland': ('Swiss', 0.5545247197151184),
 'Egypt': ('Cairo', 0.6916870474815369),
 'Australia': ('Canberra', 0.7088856101036072),
 'Vietnam': ('Hanoi', 0.5944403409957886),
 'Cuba': ('Havana', 0.6774359941482544),
 'Finland': ('Finnish', 0.6144194006919861),
 'Pakistan': ('Islamabad', 0.6850728392601013),
 'Afghanistan': ('Afghan', 0.5851746797561646),
 'England': ('Manchester', 0.5236270427703857),
 'Spain': ('Madrid', 0.6461459398269653),
 'Russia': ('Moscow', 0.7309646010398865),
 'Norway': ('Oslo', 0.599877655506134),
 'France': ('Paris', 0.6327552199363708),
 'Italy': ('Milan', 0.6697401404380798),
 'Sweden': ('Stockholm', 0.6253367066383362),
 'Iran': ('Tehran', 0.7453946471214294),
 'Japan': ('Tokyo', 0.6780205965042114)}

Use the correct answers, also given, to evaluate how accurate your capital-predictor is. You should
calculate the average accuracy over all possible training pairs.

In [56]:
cap_pred = capital_predictor('Ottawa', 'Canada')

cap_pred

{'Greece': ('Athens', 0.546576738357544),
 'Iraq': ('Baghdad', 0.6022132039070129),
 'Thailand': ('Bangkok', 0.6644393801689148),
 'China': ('Beijing', 0.6791781187057495),
 'Germany': ('Hamburg', 0.6264716386795044),
 'Switzerland': ('Swiss', 0.5545247197151184),
 'Egypt': ('Cairo', 0.6916870474815369),
 'Australia': ('Canberra', 0.7088856101036072),
 'Vietnam': ('Hanoi', 0.5944403409957886),
 'Cuba': ('Havana', 0.6774359941482544),
 'Finland': ('Finnish', 0.6144194006919861),
 'Pakistan': ('Islamabad', 0.6850728392601013),
 'Afghanistan': ('Afghan', 0.5851746797561646),
 'England': ('Manchester', 0.5236270427703857),
 'Spain': ('Madrid', 0.6461459398269653),
 'Russia': ('Moscow', 0.7309646010398865),
 'Norway': ('Oslo', 0.599877655506134),
 'France': ('Paris', 0.6327552199363708),
 'Italy': ('Milan', 0.6697401404380798),
 'Sweden': ('Stockholm', 0.6253367066383362),
 'Iran': ('Tehran', 0.7453946471214294),
 'Japan': ('Tokyo', 0.6780205965042114)}

In [84]:
capit, scorex = cap_pred['Greece']
capit

'Athens'

In [114]:
def evaluate_one_pair(train_capital, train_country):

    pairs_count = testtuples['capital-common-countries']
    predictions = capital_predictor(train_capital, train_country)

    correct = 0
    

    for true_capital, country in pairs_count:
        if country == train_country:
            continue

        pred_capital, score = predictions[country]

        if pred_capital == true_capital:
            correct += 1

    return correct / (len(pairs_count)-1)

In [115]:
evaluate_one_pair("London", "England")

0.8181818181818182

In [81]:
def average_accuracy():
    pairs_count = testtuples['capital-common-countries']
    accuracies = {}

    for train_capital, train_country in pairs_count:
        acc = evaluate_one_pair(train_capital, train_country)
        accuracies[(train_capital, train_country)] = acc

    avg_acc = sum(accuracies.values()) / len(accuracies)

    return avg_acc, accuracies

In [82]:
avg_acc, accuracies = average_accuracy()

print("Average accuracy:", avg_acc)

for pair, acc in accuracies.items():
    print(pair, "->", acc)


Average accuracy: 0.7470355731225297
('Athens', 'Greece') -> 0.8181818181818182
('Baghdad', 'Iraq') -> 0.5909090909090909
('Bangkok', 'Thailand') -> 0.7727272727272727
('Beijing', 'China') -> 0.7272727272727273
('Berlin', 'Germany') -> 0.8636363636363636
('Bern', 'Switzerland') -> 0.5454545454545454
('Cairo', 'Egypt') -> 0.8181818181818182
('Canberra', 'Australia') -> 0.7272727272727273
('Hanoi', 'Vietnam') -> 0.6363636363636364
('Havana', 'Cuba') -> 0.6818181818181818
('Helsinki', 'Finland') -> 0.8636363636363636
('Islamabad', 'Pakistan') -> 0.8181818181818182
('Kabul', 'Afghanistan') -> 0.6363636363636364
('London', 'England') -> 0.8181818181818182
('Madrid', 'Spain') -> 0.7272727272727273
('Moscow', 'Russia') -> 0.8181818181818182
('Oslo', 'Norway') -> 0.8181818181818182
('Ottawa', 'Canada') -> 0.7272727272727273
('Paris', 'France') -> 0.7727272727272727
('Rome', 'Italy') -> 0.7727272727272727
('Stockholm', 'Sweden') -> 0.8636363636363636
('Tehran', 'Iran') -> 0.5909090909090909
('T

<b>Looking at your predictions, can you think of an easy way to improve performance?</b>

Model can predicts but sometimes even the destination country itself. These are invalid answers, but the model still considers them.

##### We should filter out obvious wrong candidates.
Why this works (important)

Word2Vec sometimes ranks:

"London" very high (because it’s strongly related)
"England" also high
"Greece" itself sometimes appears

These are shortcut solutions, not real analogy reasoning.

By filtering them out, you force the model to:

actually solve the analogy

##### Why some pairs are worse:
Reason:

These capitals:

appear less frequently
have weaker embeddings
or are more ambiguous

Filtering helps especially here.



##### Deeper insight (MSc level)

We are implicitly changing:

$\hat{w} = \arg\max_{w \in V} \cos\left(\mathbf{v}_w, \mathbf{v}_{des} + \mathbf{v}_{cap} - \mathbf{v}_{country}\right)$

into:

$\hat{w} = \arg\max_{w \notin S} \cos(\cdot)$

where 

$S =$ forbidden words

👉 This is constrained decoding

Same idea used in:

beam search constraints
LLM decoding filters
structured prediction

##### Token normalisations methods
Lowercase everything etc.

###  Adapt your code so that you can predict the country of which a city is capital. Is performance the same, higher or lower this way round?

In [129]:
def  country_predictor(train_capital, train_country):
    pairs_count = testtuples['capital-common-countries']
    predictions = {}

    for true_capital, des_country in pairs_count:
        if true_capital == train_capital:
            continue

        pred = mymodel.most_similar(
            positive=[true_capital, train_country],
            negative=[train_capital],
            topn=10
        )

        predictions[true_capital] = pred

        # Filtering out
        for word, score in pred:
            if word not in {train_capital, train_country, true_capital}:
                predictions[true_capital] = (word, score)
                break

    return predictions

In [93]:
pred15 = mymodel.most_similar(positive=["Tokyo", "England"], negative=["London"], topn=10)
pred15

[('Japan', 0.6889491081237793),
 ('Shizuoka', 0.57557612657547),
 ('Japanese', 0.5434222221374512),
 ('Brave_Blossoms', 0.5234270095825195),
 ('Japans', 0.51199871301651),
 ('Iwate_Prefecture', 0.5090985298156738),
 ('Aomori_Prefecture', 0.507921040058136),
 ('defender_Tsuneyasu_Miyamoto', 0.5030457377433777),
 ('South_Korea', 0.5022144913673401),
 ('Miho_Saeki', 0.4990890622138977)]

In [130]:
country_predictor("London", "England")

{'Athens': ('Greece', 0.5293722152709961),
 'Baghdad': ('Iraqi', 0.5854012370109558),
 'Bangkok': ('Thailand', 0.6264095306396484),
 'Beijing': ('China', 0.6181187033653259),
 'Berlin': ('Germany', 0.5952338576316833),
 'Bern': ('Hanover', 0.45303669571876526),
 'Cairo': ('Egypt', 0.5641492009162903),
 'Canberra': ('Australia', 0.6054890751838684),
 'Hanoi': ('Viet_Nam', 0.5810424089431763),
 'Havana': ('Cuba', 0.5627489686012268),
 'Helsinki': ('Finland', 0.621544599533081),
 'Islamabad': ('Pakistan', 0.6654638051986694),
 'Kabul': ('Afghan', 0.5733119249343872),
 'Madrid': ('Spain', 0.6087851524353027),
 'Moscow': ('Russia', 0.6502718329429626),
 'Oslo': ('Norway', 0.5689429640769958),
 'Ottawa': ('Canada', 0.5256192088127136),
 'Paris': ('France', 0.6676377654075623),
 'Rome': ('Italy', 0.5715967416763306),
 'Stockholm': ('Sweden', 0.6610991358757019),
 'Tehran': ('Iran', 0.6443779468536377),
 'Tokyo': ('Japan', 0.6889491081237793)}

In [100]:
for country, score in aa['Athens']:
    if country == "Greece":
    
        print(score)        

0.5293722152709961


In [131]:
def evaluate_country_pair(train_capital, train_country):

    pairs_count = testtuples['capital-common-countries']
    predictions = country_predictor(train_capital, train_country)

    
    correct = 0
    total = 0
    for true_capital, true_country in pairs_count:
        if true_capital == train_capital:
            continue

        
        pred_country, score = predictions[true_capital]

        if pred_country == true_country:
            correct += 1

        total += 1

    return correct / total

In [132]:
evaluate_country_pair('London', 'England')

0.8181818181818182

In [133]:
def average_accuracy():
    pairs_count = testtuples['capital-common-countries']
    accuracies = {}

    for train_capital, train_country in pairs_count:
        acc = evaluate_country_pair(train_capital, train_country)
        accuracies[(train_capital, train_country)] = acc

    avg_acc = sum(accuracies.values()) / len(accuracies)

    return avg_acc, accuracies

In [134]:
avg_acc, accuracies = average_accuracy()

print("Average accuracy:", avg_acc)

for pair, acc in accuracies.items():
    print(pair, "->", acc)

Average accuracy: 0.8320158102766798
('Athens', 'Greece') -> 0.8181818181818182
('Baghdad', 'Iraq') -> 0.8181818181818182
('Bangkok', 'Thailand') -> 0.8636363636363636
('Beijing', 'China') -> 0.7727272727272727
('Berlin', 'Germany') -> 0.8181818181818182
('Bern', 'Switzerland') -> 0.8636363636363636
('Cairo', 'Egypt') -> 0.8181818181818182
('Canberra', 'Australia') -> 0.8181818181818182
('Hanoi', 'Vietnam') -> 0.9090909090909091
('Havana', 'Cuba') -> 0.9090909090909091
('Helsinki', 'Finland') -> 0.8181818181818182
('Islamabad', 'Pakistan') -> 0.8181818181818182
('Kabul', 'Afghanistan') -> 0.8181818181818182
('London', 'England') -> 0.8181818181818182
('Madrid', 'Spain') -> 0.8181818181818182
('Moscow', 'Russia') -> 0.8181818181818182
('Oslo', 'Norway') -> 0.8181818181818182
('Ottawa', 'Canada') -> 0.8181818181818182
('Paris', 'France') -> 0.8181818181818182
('Rome', 'Italy') -> 0.8181818181818182
('Stockholm', 'Sweden') -> 0.8181818181818182
('Tehran', 'Iran') -> 0.8181818181818182
('T

Is performance same, higher, or lower?

In practice, it is often lower when predicting country from capital.

Why?

- capitals are often strongly associated with their country, but
- country vectors are usually broader and noisier semantic objects
- some capital vectors are cleaner landmarks than country vectors
- embeddings often encode “Athens → Greece” less symmetrically than “Greece → Athens”

Also country names have more varied contexts:

- politics
- people
- language
- culture
- demonyms

while capitals often behave more like named entities.

So usually:

- capital prediction from country performs better
- country prediction from capital performs a bit worse

But you should confirm with your actual model output.

Key takeaway

So: your analogy formula is basically fine, but your evaluation metric is not accuracy.

Replace score-averaging with exact-match top-1 accuracy. Then compare the average country-prediction accuracy against your earlier capital-prediction accuracy of about 0.747.

### Adapt your code so that you can consider any of the relationships in testtuples. Rank the relationships in order of easiness to predict. Why do you think some are easier than others?